test/validation/api/ipsec/ipsec.c

```diff
@@ -29,16 +29,28 @@ static odp_pktio_t pktio_create(odp_pool_t pool)
 	odp_pktio_t pktio;
 	odp_pktio_param_t pktio_param;
 	odp_pktin_queue_param_t pktin_param;
-	odp_pktio_capability_t capa;
+	odp_pktio_capability_t pktio_capa;
+	odp_ipsec_capability_t ipsec_capa;
 
 	int ret;
 
 	if (pool == ODP_POOL_INVALID)
 		return ODP_PKTIO_INVALID;
 
+	if (odp_ipsec_capability(&ipsec_capa) < 0) {
+		fprintf(stderr, "IPsec capabilities failed.\n");
+		return ODP_PKTIO_INVALID;
+	}
+
 	odp_pktio_param_init(&pktio_param);
 	pktio_param.in_mode = ODP_PKTIN_MODE_QUEUE;
 
+	if (suite_context.inbound_op_mode == ODP_IPSEC_OP_MODE_INLINE ||
+	    suite_context.inbound_op_mode == ODP_IPSEC_OP_MODE_INLINE) {
+		if (!ipsec_capa.queue_type_plain && ipsec_capa.queue_type_sched)
+			pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
```

Comment:
This cannot be right as the two or'ed expressions are the same.

Actually, I do not see why the pktio queue type should be affected by the IPsec capability that concerns IPsec completion queues.

Reply:
This was a typo. I meant to set in_mode to sched if we are running inline mode test. Thank you spotting this.

On our platform, with inline IPsec, we can submit only via scheduled queues and that would be the pktin queues configured for the packet_io device.

Reply:
> On our platform, with inline IPsec, we can submit only via scheduled queues and that would be the pktin queues configured for the packet_io device.

API validation tests are written against the API spec, not any particular implementation.

Setting pktio queue type in this patch is wrong. This patch is about checking _ipsec_ destination queue capability, which does not have anything (in the API spec) to do with pktio queue type.

As discussed, your implementation is currently not API compliant with respect to the destination queue in inline mode. For this we need an API spec change that allows inline processed packets to be delivered through pktio queues. And for that API change we need to change the validation test accordingly. But at this point, this pktio queue type change looks totally unexpected to anyone who does not know the backround and what future API changes might be coming.

Btw, are you saying that in your implementation there it not only the limitation that inline processed packets can only be delivered through pktio queueus, but also the limitation that said pktio queues cannot be plain queues? But you do support plain pktio queues without IPsec do you?

Reply:
Removed the pktin queue configuration introduced inside pktio_create() in v9. Updated the pktio_create() function to check whether inline mode is supported using the pktio capability.